torch/nn/_functions/rnn.py

```diff
@@ -7,19 +7,22 @@
     pass
 
 
-def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.relu(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None, skip_input=False):
+    xw_ih = input if skip_input else F.linear(input, w_ih, b_ih)
+    hy = F.relu(xw_ih + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
-def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.tanh(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None, skip_input=False):
+    xw_ih = input if skip_input else F.linear(input, w_ih, b_ih)
+    hy = F.tanh(xw_ih + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
-def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
+def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None, skip_input=False):
     hx, cx = hidden
-    gates = F.linear(input, w_ih, b_ih) + F.linear(hx, w_hh, b_hh)
+    xw_ih = input.repeat(1, 4) if skip_input else F.linear(input, w_ih, b_ih)
```

Comment:
Making that change for LSTM and GRU cell I had to squeeze the additional dim like below:

```
def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
    hx, cx = hidden
    x_h = input.unsqueeze(1).expand(input.size(0), 4, input.size(1)) if w_ih is None else F.linear(input, w_ih, b_ih)
    gates = x_h + F.linear(hx, w_hh, b_hh)
    ingate, forgetgate, cellgate, outgate = [x.squeeze(1) for x in gates.chunk(4, 1)] # Remove single dim

    ingate = F.sigmoid(ingate)
    forgetgate = F.sigmoid(forgetgate)
    cellgate = F.tanh(cellgate)
    outgate = F.sigmoid(outgate)

    cy = (forgetgate * cx) + (ingate * cellgate)
    hy = outgate * F.tanh(cy)

    return hy, cy
```

This works for GRUs, however causing an error in the backprop:

```
Traceback (most recent call last):
  File "/home/sean/test/problem.py", line 89, in test
    loss.backward()
  File "/home/sean/anaconda3/lib/python3.5/site-packages/torch/autograd/variable.py", line 145, in backward
    self._execution_engine.run_backward((self,), (gradient,), retain_variables)
  File "/home/sean/anaconda3/lib/python3.5/site-packages/torch/nn/_functions/linear.py", line 22, in backward
    grad_input = torch.mm(grad_output, weight)
RuntimeError: matrices expected, got 3D, 2D tensors at /home/sean/Work/pytorch/torch/lib/TH/generic/THTensorMath.c:857
```

Any ideas?